libvast/src/type.cpp

```diff
@@ -238,7 +238,55 @@ type::type(std::string_view name, const type& nested,
     // types, which did not require an legacy alias type wrapping to have a name.
     *this = nested;
   } else {
-    const auto nested_bytes = as_bytes(nested);
+    auto nested_bytes = as_bytes(nested);
+    // Identify the first named metadata-layer, and store all attributes we
+    // encounter until then. We merge the attributes into the attributes
+    // provided to this constructor, priorising the new attributes, with the
+    // nested byte range being adjusted to the first named metadata-layer (or
+    // the underlying concrete type).
+    auto merged_attributes = std::vector<attribute_view>{};
+    for (const auto* root = fbs::GetType(nested_bytes.data());
+         root != nullptr;) {
+      switch (root->type_type()) {
+        case fbs::type::Type::NONE:
+        case fbs::type::Type::bool_type:
+        case fbs::type::Type::integer_type:
+        case fbs::type::Type::count_type:
+        case fbs::type::Type::real_type:
+        case fbs::type::Type::duration_type:
+        case fbs::type::Type::time_type:
+        case fbs::type::Type::string_type:
+        case fbs::type::Type::pattern_type:
+        case fbs::type::Type::address_type:
+        case fbs::type::Type::subnet_type:
+        case fbs::type::Type::enumeration_type:
+        case fbs::type::Type::list_type:
+        case fbs::type::Type::map_type:
+        case fbs::type::Type::record_type:
+          root = nullptr;
+          break;
+        case fbs::type::Type::enriched_type: {
+          const auto* enriched_type = root->type_as_enriched_type();
+          if (enriched_type->name()) {
+            root = nullptr;
+            break;
+          }
+          if (const auto* attributes = enriched_type->attributes()) {
+            for (const auto* attribute : *attributes) {
+              VAST_ASSERT(attribute->key());
+              merged_attributes.push_back({
+                attribute->key()->string_view(),
+                attribute->value() ? attribute->value()->string_view() : "",
+              });
+            }
+          }
+          nested_bytes = as_bytes(*enriched_type->type());
```

Comment:
nit: This introduces an additional error into the `reserved_size` pre-calculation.

Reply:
I don't see how. Can you elaborate? We only pre-calculate that if we do not have attributes in any form.

Reply:
We talked about this in a call. The code is correct here.